dropshot/src/pagination.rs

```diff
@@ -0,0 +1,569 @@
+pub enum MarkerScanMode {
+    ByMarkerAsc,
+}
+
+/**
+ * Internal type used as the `PageSelector` type for `MarkerPaginator`.
+ */
+#[derive(Debug, Deserialize, ExtractedParameter, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum MarkerPageSelector<F> {
+    Marker(F),
+}
+
+/*
+ * Generic pagination infrastructure
+ * XXX need to better understand the trait bounds / serde bounds here.
+ */
+
+/**
+ * Query parameters provided by clients when scanning a paginated collection
+ *
+ * See `PaginatedResource` for more information.
+ */
+#[derive(Debug, Deserialize, ExtractedParameter)]
+#[serde(bound(
+    deserialize = "PageSelector: DeserializeOwned, ScanMode: DeserializeOwned"
+))]
+pub struct PaginationParams<ScanMode, PageSelector>
+where
+    ScanMode: Debug + Send + Sync + 'static,
+    PageSelector: Serialize + Send + Sync + 'static,
+{
+    /**
+     * Specifies either how the client wants to begin a new scan or how to
+     * resume a previous scan.  This field is flattened by serde, so see the
+     * variants of [`WhichPage`] to see which fields actually show up in the
+     * serialized form.
+     */
+    #[serde(flatten)]
+    pub page_params: WhichPage<ScanMode, PageSelector>,
+
+    /**
+     * Optional client-requested limit on page size.  Consumers should use
+     * [`dropshot::RequestContext::page_limit()`] to access this value.
+     */
+    pub(crate) limit: Option<NonZeroU64>,
+}
+
+/**
+ * Indicates whether the client is beginning a new scan or resuming an existing
+ * one and provides the corresponding parameters for each case
+ */
+/*
+ * In REST APIs, callers typically provide either the parameters to resume a
+ * scan or the parameters to begin a new one, with no separate field indicating
+ * which case they're requesting.  Fortunately, serde(untagged) implements this
+ * behavior precisely, with one caveat: the variants below are tried in order
+ * until one succeeds.  So for this to work, `NextPage` must be defined before
+ * `FirstPage`, since any set of parameters at all (including no parameters at
+ * all) are valid for `FirstPage`.
+ */
+#[derive(Debug, Deserialize, ExtractedParameter)]
+#[serde(bound(
+    deserialize = "PageSelector: DeserializeOwned, ScanMode: DeserializeOwned"
+))]
+#[serde(untagged)]
+pub enum WhichPage<ScanMode, PageSelector>
+where
+    ScanMode: Debug + Send + Sync + 'static,
+    PageSelector: Serialize + Send + Sync + 'static,
+{
+    /**
+     * Indicates that the client is resuming a previous scan.  The client
+     * provides `page_token`, an opaque value that wraps the consumer's
+     * `PageSelector` type.  Critically, this must provide enough information
+     * for the consumer to resume a scan.
+     */
+    NextPage { page_token: PageToken<PageSelector> },
+
+    /**
+     * Indicates that the client is beginning a new scan.  The client may
+     * provide `list_mode`, a consumer-specified type.
+     */
+    FirstPage { list_mode: Option<ScanMode> },
```

Comment:
I don't have an opinion, but why did you choose to have these as named elements rather than the tuple form i.e.
```rust
NextPage(PageToken<PageSelector>),
FirstPage(Option<ScanMode>),
```
Is it to affect the encoding?

Reply:
Yes.  This is how `list_mode` winds up being the name of the query parameter.  I don't think it's important for the `NextPage` variant because of the `#[serde(flatten)]`.

Related to this: the way these structures are set up feels a bit elaborate because of the way we're building on `Query` + serde to do all of the deserialization.  The requirements to make the query params feel idiomatic (i.e. `list_mode` or `page_token`) results in having this untagged serde enum where the order of fields matters, and it also results in consumers dealing with this extra `page_token` field when they pull the value out of `NextPage`, and that thing uses `deserialize_from`.

Another option would be to do more of this explicitly instead of relying on all these serde attributes.  Specifically, we could modify `Query` so that instead of taking a `Deserialize`, it consumes a new trait `FromQuerystring`.  Then we could `impl<T> FromQueryString for T: Deserialize` to cover all of the existing use cases today that use `Deserialize`.  For `PaginationParams`, we could implement a `FromQuerystring` that explicitly parses the string and fills in the structures we want.  This could let us change these variants to use the tuple form, which would be more concise when destructuring it, as well as eliminate the `page_token` indirection from the interface.

Reply:
I see. While it leads to some ugly cases, it does seem like a nice short cut in terms of implementation *and* education to lean on serde here regardless.

Reply:
FWIW, I ended up finding a middle ground here.  I created `RawPaginationParams` and `RawWhichPage` which use the serde trickery that we were doing before to parse stuff right off the wire.  Then I've implemented `PaginationParams` in a way that causes serde to first parse it as a `RawPaginationParams` and then use `TryFrom` to convert it into a `PaginationParams`.  This way, we still use the serde trickery that we had previously (so we don't have to implement `Deserialize` by hand), but we can expose a much cleaner structure to consumers.  More on this in another comment.